### PR TITLE
Add EndpointPrefix Support

### LINKED ISSUE #197: Support the Smithy Endpoint Trait
The involves endpoint prefix support. WIP branch: `endpoint-prefix`

### PR DESCRIPTION
**Pull request by rcoh**
*Issue #, if available:* Fixes #197 
*Description of changes:* This diff finalizes our endpoint support by adding support for the endpointPrefix trait.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: codegen/src/test/kotlin/software/amazon/smithy/rust/codegen/smithy/EndpointTraitBindingsTest.kt [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        val (ctx, testDir) = generatePluginContext(model, "com.example#TestService")
+        val testDecorator = object : RustCodegenDecorator {
+            override val name: String = "Test"
+            override val order: Byte = 0
+            override fun extras(protocolConfig: ProtocolConfig, rustCrate: RustCrate) {
+                rustCrate.withModule(RustModule.default("test", public = false)) {
+                    it.unitTest(
+                        """
+                let conf = crate::Config::builder().build();
+                crate::operation::SayHello::builder().greeting("hey there!").build(&conf).expect_err("no spaces or exclamation points in ep prefixes");
```
Comment (jdisanti2019) [inline codegen/src/test/kotlin/software/amazon/smithy/rust/codegen/smithy/EndpointTraitBindingsTest.kt:125]: Should there also be a test for empty host labels (from the Client Behavior section of the spec)?

> Clients MUST fail when expanding a hostPrefix template if the value of any labeled member is empty or null.

Reference: https://awslabs.github.io/smithy/1.0/spec/core/endpoint-traits.html#client-behavior

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/EndpointTraitBindings.kt [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            writer.rust("#T::endpoint::EndpointPrefix::new(${endpointTrait.prefixFormatString()})", smithyHttp)
+        } else {
+            writer.withBlock(
+                "#T::endpoint::EndpointPrefix::new(format!(${endpointTrait.prefixFormatString()}, ",
+                "))",
+                smithyHttp
+            ) {
+                endpointTrait.hostPrefix.labels.forEach { label ->
+                    val member = inputShape.getMember(label.content).get()
+                    rust("${label.content} = $input.${symbolProvider.toMemberName(member)}, ")
```
Comment (jdisanti2019) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/EndpointTraitBindings.kt:49]: For my understanding, how does validation end up running on these label values? Is it ultimately the final endpoint validation that catches bad values for this?

Comment (rcoh) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/EndpointTraitBindings.kt:49]: yeah there isn't any validators in the builders

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/SymbolVisitor.kt [hunk 3]
```diff
@@ -137,10 +144,14 @@ class SymbolVisitor(
...
     private fun handleOptionality(symbol: Symbol, member: MemberShape, container: Shape): Symbol {
-        // If a field has the httpLabel trait and we are generating
+        // If a field has the httpLabel or hostLabel trait and we are generating
         // an Input shape, then the field is _not optional_.
         val httpLabeledInput =
-            container.hasTrait(SyntheticInputTrait::class.java) && member.hasTrait(HttpLabelTrait::class.java)
+            (container.hasTrait(SyntheticInputTrait::class.java) || container.hasTrait(InputBodyTrait::class.java)) && (
+                member.hasTrait(HttpLabelTrait::class.java) || member.hasTrait(
+                    HostLabelTrait::class.java
+                )
```
[the remark this answers is not in the document]
Comment (aajtodd) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/SymbolVisitor.kt:153]: We defined many of these already FYI (has/expect/get/etc), feel free to steal them (`model/ShapeExt.kt`)

```kt
/**
 * Kotlin sugar for hasTrait() check. e.g. shape.hasTrait<EnumTrait>() instead of shape.hasTrait(EnumTrait::class.java)
 */
inline fun <reified T : Trait> Shape.hasTrait(): Boolean = hasTrait(T::class.java)
```

Separate PR of course

[the remark this answers is not in the document]
Comment (jdisanti2019) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/SymbolVisitor.kt:153]: Refactored in #385

**Comment by rcoh** [comment]
landed separately
